File: server/data/test-project/java-service/src/main/java/com/testproject/service/UserService.java
package com.testproject.service;

import com.testproject.model.User;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import javax.persistence.EntityManager;
import javax.persistence.Query;
import java.util.List;
import java.util.Optional;

/**
 * User Service
 * Business logic for user operations
 * 
 * BUG: Contains SQL Injection vulnerability in searchUsers method
 */
@Service
public class UserService {

    @Autowired
    private EntityManager entityManager;

    /**
     * Search users by name or email
     * 
     * SECURITY BUG: SQL Injection vulnerability!
     * The search parameter is concatenated directly into the SQL query
     * without proper parameterization.
     * 
     * Example attack: search = "'; DROP TABLE users; --"
     */
    public List<User> searchUsers(String search) {
        // BUG: Raw string concatenation - vulnerable to SQL injection
        String sql = "SELECT * FROM users WHERE name LIKE '%" + search + "%' " +
                    "OR email LIKE '%" + search + "%'";
        
        Query query = entityManager.createNativeQuery(sql, User.class);
        return query.getResultList();
    }

    /**
     * Get all users with pagination (safe implementation)
     */
    public List<User> getAllUsers(int page, int limit) {
        String jpql = "SELECT u FROM User u ORDER BY u.createdAt DESC";
        return entityManager.createQuery(jpql, User.class)
                .setFirstResult((page - 1) * limit)
                .setMaxResults(limit)
                .getResultList();
    }

    /**
     * Get user by ID
     */
    public Optional<User> getUserById(Long id) {
        User user = entityManager.find(User.class, id);
        return Optional.ofNullable(user);
    }

    /**
     * Create new user
     */
    public User createUser(User user) {
        entityManager.persist(user);
        return user;
    }

    /**
     * Update user
     */
    public Optional<User> updateUser(Long id, User userData) {
        User existing = entityManager.find(User.class, id);
        if (existing == null) {
            return Optional.empty();
        }
        
        existing.setName(userData.getName());
        existing.setEmail(userData.getEmail());
        entityManager.merge(existing);
        
        return Optional.of(existing);
    }

    /**
     * Delete user
     */
    public void deleteUser(Long id) {
        User user = entityManager.find(User.class, id);
        if (user != null) {
            entityManager.remove(user);
        }
    }

    /**
     * SECURE version of search (for reference)
     * This is how the searchUsers method should be implemented
     */
    public List<User> searchUsersSecure(String search) {
        String jpql = "SELECT u FROM User u WHERE u.name LIKE :search OR u.email LIKE :search";
        return entityManager.createQuery(jpql, User.class)
                .setParameter("search", "%" + search + "%")
                .getResultList();
    }
}

File: server/data/test-project/node-api/src/services/user.service.ts
/**
 * User Service
 * Business logic for user management
 */

import { dbService } from './db.service';
import { User } from '../types';

interface ListUsersOptions {
  page: number;
  limit: number;
  search?: string;
}

interface PaginatedResult<T> {
  data: T[];
  pagination: {
    page: number;
    limit: number;
    total: number;
    totalPages: number;
  };
}

class UserService {
  /**
   * List users with pagination
   */
  async listUsers(options: ListUsersOptions): Promise<PaginatedResult<Omit<User, 'password'>>> {
    const { page, limit, search } = options;
    const offset = (page - 1) * limit;
    
    let query = 'SELECT id, email, name, created_at, is_admin FROM users WHERE 1=1';
    let countQuery = 'SELECT COUNT(*) as count FROM users WHERE 1=1';
    const params: any[] = [];
    
    if (search) {
      params.push(`%${search}%`);
      query += ` AND (email ILIKE $${params.length} OR name ILIKE $${params.length})`;
      countQuery += ` AND (email ILIKE $${params.length} OR name ILIKE $${params.length})`;
    }
    
    query += ` ORDER BY created_at DESC LIMIT $${params.length + 1} OFFSET $${params.length + 2}`;
    params.push(limit, offset);
    
    const [users, countResult] = await Promise.all([
      dbService.query<Omit<User, 'password'>>(query, params),
      dbService.query<{ count: string }>(countQuery, params.slice(0, -2)),
    ]);
    
    const total = parseInt(countResult[0]?.count || '0', 10);
    
    return {
      data: users,
      pagination: {
        page,
        limit,
        total,
        totalPages: Math.ceil(total / limit),
      },
    };
  }

  /**
   * Get user by ID
   */
  async getUserById(id: string): Promise<User | null> {
    const users = await dbService.query<User>(
      'SELECT * FROM users WHERE id = $1',
      [id]
    );
    return users[0] || null;
  }

  /**
   * Update user
   */
  async updateUser(id: string, data: { name?: string; email?: string }): Promise<User | null> {
    const result = await dbService.query<User>(
      `UPDATE users 
       SET name = COALESCE($1, name),
           email = COALESCE($2, email),
           updated_at = NOW()
       WHERE id = $3
       RETURNING *`,
      [data.name, data.email, id]
    );
    return result[0] || null;
  }

  /**
   * Delete user
   */
  async deleteUser(id: string): Promise<void> {
    await dbService.query('DELETE FROM users WHERE id = $1', [id]);
  }

  /**
   * Check if user is admin
   */
  async isAdmin(id: string): Promise<boolean> {
    const result = await dbService.query<{ is_admin: boolean }>(
      'SELECT is_admin FROM users WHERE id = $1',
      [id]
    );
    return result[0]?.is_admin || false;
  }
}

export const userService = new UserService();
export default userService;

